endpoints/put.go

```diff
@@ -9,128 +9,99 @@ import (
-				// for example... so we'll need to un-escape it before we consider it to be XML content.
-				var interpreted string
-				json.Unmarshal(p.Value, &interpreted)
-				toCache = p.Type + interpreted
-			} else if p.Type == backends.JSON_PREFIX {
-				toCache = p.Type + string(p.Value)
-			} else {
-				http.Error(w, fmt.Sprintf("Type must be one of [\"json\", \"xml\"]. Found %v", p.Type), http.StatusBadRequest)
-				return
-			}
-
-			logrus.Debugf("Storing value: %s", toCache)
-			u2, err := uuid.NewV4()
-			if err != nil {
-				http.Error(w, fmt.Sprintf("Error generating version 4 UUID"), http.StatusInternalServerError)
-			}
-			resps.Responses[i].UUID = u2.String()
-
-			ctx, cancel := context.WithTimeout(context.Background(), 500*time.Millisecond)
-			defer cancel()
-			// Only allow setting a provided key if configured (and ensure a key is provided).
-			if allowKeys && len(p.Key) > 0 {
-				s, err := backend.Get(ctx, p.Key)
-				if err != nil || len(s) == 0 {
-					resps.Responses[i].UUID = p.Key
+		// Get a response object from the resource pool that we'll fill with processed info
+		backend.resp = putResponsePool.Get().(*PutResponse)
+		backend.resp.Responses = make([]PutResponseObject, len(backend.put.Puts))
+		backend.resp.toCacheStrings = make([]string, len(backend.put.Puts))
+		defer putResponsePool.Put(backend.resp)
+
+		var cancel func()
+		backend.ctx, cancel = context.WithTimeout(context.Background(), 500*time.Millisecond)
+		defer cancel()
+
+		exitData := &exitInfo{errMsg: "", status: http.StatusOK}
+
+		if len(backend.put.Puts) == 0 {
+			exitData.errMsg = "No keys were sent in backend request"
+			exitData.status = http.StatusBadRequest
+		} else if len(backend.put.Puts) > maxNumValues {
+			exitData.errMsg = fmt.Sprintf("More keys than allowed: %d", maxNumValues)
+			exitData.status = http.StatusBadRequest
+		} else { // 1 <= len(backend.put.Puts) < maxNumValues
+			// Processonly if all requests come error free
+			validateAndEncode(backend.put, backend.resp, exitData)
+			if exitData.status == http.StatusOK {
+				if len(backend.put.Puts) == 1 {
+					callBackendGet(backend, 0)
+					callBackendPut(backend, exitData, 0)
 				} else {
-					resps.Responses[i].UUID = ""
-				}
-			}
-			// If we have a blank UUID, don't store anything.
-			// Eventually we may want to provide error details, but as of today this is the only non-fatal error
-			// Future error details could go into a second property of the Responses object, such as "errors"
-			if len(resps.Responses[i].UUID) > 0 {
-				err = backend.Put(ctx, resps.Responses[i].UUID, toCache, p.TTLSeconds)
-				if err != nil {
-					if _, ok := err.(*backendDecorators.BadPayloadSize); ok {
-						http.Error(w, fmt.Sprintf("POST /cache element %d exceeded max size: %v", i, err), http.StatusBadRequest)
-						return
+					// Run process in parallel
+					getIndexChannel := make(chan int)
```

Comment:
A single channel should be enough. Each go routine running in parallel needs to 1) possibly make a get call to the backend (to verify a user provided key is unique.. although that's an issue for another day) and then 2) always make a put call to the backend.

Since steps 1 and 2 always must be synchronous within each parallelized put, I don't see a reason to separate those actions and take additional overhead cost. Kicking off all go routines and waiting for them to complete should suffice.

Taking this a step further, we don't need to use channels at all. Each concurrent goroutine can store their result in the array at a predetermined index, as you have in this code, so that shouldn't cause any memory issues. You can use a wait group to coordinate completion instead, which should be a bit faster.

Reply:
Done. USed `WaitGroup` instead of channels and the code really looks cleaner